packages/ses/src/make-safe-evaluator.js

```diff
@@ -0,0 +1,107 @@
+// Portions adapted from V8 - Copyright 2016 the V8 project authors.
+// https://github.com/v8/v8/blob/master/src/builtins/builtins-function.cc
+
+import {
+  WeakSet,
+  apply,
+  immutableObject,
+  proxyRevocable,
+  weaksetAdd,
+} from './commons.js';
+import { getScopeConstants } from './scope-constants.js';
+import { createScopeHandler } from './scope-handler.js';
+import { applyTransforms, mandatoryTransforms } from './transforms.js';
+import { makeEvaluateFactory } from './make-evaluate-factory.js';
+import { assert } from './error/assert.js';
+
+const { details: d } = assert;
+
+// TODO: rename localObject to scopeObject
+
+/**
+ * makeSafeEvaluator()
+ * Build the low-level operation used by all evaluators:
+ * eval(), Function(), Compartment.prototype.evaluate().
+ *
+ * @param {Object} options
+ * @param {Object} options.globalObject
+ * @param {Object} [options.localObject]
+ * @param {Array<Transform>} [options.globalTransforms]
+ * @param {bool} [options.sloppyGlobalsMode]
+ * @param {WeakSet} [options.knownScopeProxies]
+ */
+export const makeSafeEvaluator = ({
+  globalObject,
+  localObject = {},
+  globalTransforms = [],
+  sloppyGlobalsMode = false,
+  knownScopeProxies = new WeakSet(),
+} = {}) => {
+  const {
+    scopeHandler,
+    admitOneUnsafeEvalNext,
+    resetOneUnsafeEvalNext,
+  } = createScopeHandler(globalObject, localObject, {
+    sloppyGlobalsMode,
+  });
+  const { proxy: scopeProxy, revoke: revokeScopeProxy } = proxyRevocable(
+    immutableObject,
+    scopeHandler,
+  );
+  weaksetAdd(knownScopeProxies, scopeProxy);
+
+  const constants = getScopeConstants(globalObject, localObject);
+  const evaluateFactory = makeEvaluateFactory(constants);
+  const evaluate = apply(evaluateFactory, scopeProxy, []);
+
+  /**
+   * @param {string} source
+   * @param {Object} [options]
+   * @param {Array<Transform>} [options.localTransforms]
+   */
+  const safeEvaluate = (source, { localTransforms = [] } = {}) => {
```

Comment:
For readability, let’s name the intermediate `options` argument and unpack it in the body of the function.